vendor/github.com/insolar/insconfig/configurator.go

```diff
@@ -0,0 +1,236 @@
+type ConfigStruct interface {
+	GetConfig() interface{}
+}
+
+type Params struct {
+	// Your config
+	ConfigStruct ConfigStruct
+	// Prefix for environment variables
+	EnvPrefix string
+	// Custom viper decoding hooks
+	ViperHooks []mapstructure.DecodeHookFunc
+	// Should return config path
+	ConfigPathGetter ConfigPathGetter
+	// If set then return error on file not found
+	FileRequired bool
+}
+
+type ConfigPathGetter interface {
+	GetConfigPath() string
+}
+
+// Default behaviour adds "--config" flag and read path from it, custom flags should be created before
+type DefaultConfigPathGetter struct {
+	// For go flags compatibility
+	GoFlags *goflag.FlagSet
+	// For spf13/pflags compatibility
+	PFlags *flag.FlagSet
+}
+
+func (g DefaultConfigPathGetter) GetConfigPath() string {
+	if g.GoFlags != nil {
+		flag.CommandLine.AddGoFlagSet(g.GoFlags)
+	}
+	if g.PFlags != nil {
+		flag.CommandLine.AddFlagSet(g.PFlags)
+	}
+	configPath := flag.String("config", "", "path to config")
+	flag.Parse()
+	return *configPath
+}
+
+type insConfigurator struct {
+	params Params
+	viper  *viper.Viper
+}
+
+func NewInsConfigurator(params Params) insConfigurator {
+	return insConfigurator{
+		params: params,
+		viper:  viper.New(),
+	}
+}
+
+// Loads configuration from path and making checks
+func (i *insConfigurator) Load() (ConfigStruct, error) {
+	if i.params.EnvPrefix == "" {
+		return nil, errors.New("EnvPrefix should be defined")
+	}
+
+	configPath := i.params.ConfigPathGetter.GetConfigPath()
+	return i.load(configPath, i.params.FileRequired)
+}
+
+func (i *insConfigurator) load(path string, required bool) (ConfigStruct, error) {
+
+	i.viper.AutomaticEnv()
+	i.viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
+	i.viper.SetEnvPrefix(i.params.EnvPrefix)
+
+	i.viper.SetConfigFile(path)
+	if err := i.viper.ReadInConfig(); err != nil {
+		if required {
+			return nil, err
+		}
+		fmt.Printf("failed to load config from '%s'\n", path)
+	}
+	actual := i.params.ConfigStruct.GetConfig()
+	i.params.ViperHooks = append(i.params.ViperHooks, mapstructure.StringToTimeDurationHookFunc(), mapstructure.StringToSliceHookFunc(","))
+	err := i.viper.UnmarshalExact(actual, viper.DecodeHook(mapstructure.ComposeDecodeHookFunc(
+		i.params.ViperHooks...,
+	)))
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to unmarshal config file into configuration structure")
+	}
+	configStructKeys, err := i.checkAllValuesIsSet()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := i.checkNoExtraENVValues(configStructKeys); err != nil {
+		return nil, err
+	}
+
+	// Second Unmarshal needed because of bug https://github.com/spf13/viper/issues/761
+	// This should be evaluated after manual values overriding is done
+	err = i.viper.UnmarshalExact(actual, viper.DecodeHook(mapstructure.ComposeDecodeHookFunc(
+		i.params.ViperHooks...,
+	)))
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to unmarshal config file into configuration structure 2")
```

Comment:
why we need  2 here? 
is it possible to write a different suck as `failed to unmarshal the config file into configuration structure the second time`?

Reply:
No matter what is the error text, because this is imposible situation